[6.17] Display/GPU fixes

minlexx opened the pull request:
Grab description from one of:
* https://patchwork.kernel.org/project/linux-arm-msm/list/?series=1004174&state=%2A&archive=both
* https://patchwork.kernel.org/project/linux-arm-msm/list/?series=1004250&state=%2A&archive=both
* https://patchwork.kernel.org/project/linux-arm-msm/list/?series=1002506&state=%2A&archive=both

Important bit is
> This also fixes display init on Linux v6.17.

For the reference, without `MDSS_BCR` during boot display doesn't come up (black) until you turn it off/on, or you'll see plain blue display instead of kernel logs or splash screen and this will be in dmesg:
```
[  115.477507] hw recovery is not complete for ctl:2
[  115.479056] [drm:dpu_encoder_phys_vid_prepare_for_kickoff:569] [dpu error]enc33 intf1 ctl 2 reset failure: -22
[  116.181469] [drm:dpu_encoder_frame_done_timeout:2727] [dpu error]enc33 frame done timeout
[  116.345396] [drm:dpu_encoder_frame_done_timeout:2727] [dpu error]enc33 frame done timeout
[  116.748773] [drm:dpu_encoder_frame_done_timeout:2727] [dpu error]enc33 frame done timeout
[  117.213569] [drm:dpu_encoder_frame_done_timeout:2727] [dpu error]enc33 frame done timeout
[  117.645205] hw recovery is not complete for ctl:2
[  117.647726] [drm:dpu_encoder_phys_vid_prepare_for_kickoff:569] [dpu error]enc33 intf1 ctl 2 reset failure: -22
[  117.650885] [drm:dpu_encoder_frame_done_timeout:2727] [dpu error]enc33 frame done timeout
[  118.284506] [drm:dpu_encoder_frame_done_timeout:2727] [dpu error]enc33 frame done timeout
...
[  129.115542] [drm:dpu_encoder_frame_done_timeout:2727] [dpu error]enc33 frame done timeout
[  141.547007] dpu_encoder_frame_done_timeout: 2 callbacks suppressed
```
---
Also partially undo https://patchwork.freedesktop.org/series/114004/

Suspicion is that msm_gpu_fault_crashstate_capture() does something
that causes hang for the whole device, so you cannot get any crash
dumps. Faults still happen sometimes, and they are logged, but
device remains usable.

This requires running with `FD_MESA_DEBUG=noblit,inorder,gmem` otherwise your device is unusable brick of plastic.

minlexx committed:
HACK adreno: make gpu to NOT freeze whole device on fault

Suspicion is that msm_gpu_fault_crashstate_capture() does something
that causes hang for the whole device, so you cannot get any crash
dumps. Faults still happen sometimes, and they are logged, but
device remains usable.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
HACK dts: sdm660: Disable all GPU freqs above nominal for stability

Remove all GPU clock frequences that require power domain levels
above nominal to make GPU less crashy. This results in less frequent
hangchecks and more stability.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

setotau commented:
Checked on clover-plus with FD_MESA_DEBUG=noblit. GPU catches faults, but continues working.

minlexx commented:
<img width="1344" height="296" alt="image" src="https://github.com/user-attachments/assets/8c9d9769-b6fa-4b88-be83-4c726a770e05" />

Soo gpu iommu faults do not kill device completely now, which is an improvement.  These `*** gpu fault:` messages which are new on 6.17 and make device completely hang killing SSH-USB connection immediately are now non-fatal.

but "timeout to drain ringbuffer"/"hangcheck recover" make device unusable as it always was before (not a regression)

minlexx commented:
I only tested it on lavender so far. Maybe similar trim to gpucc frequencies is also required for sdm630 too. Need to test on more devices before proceeding

ZXlieC commented:
Here i have completely solve this problem: https://github.com/ZXlieC/xec-mainline-project/commit/5e527d335f4becbd863d3561953ff87bdcac2e29

I have tested kernel on xiaomi-lavender with Arch Linux ARM and pmos and it works normaly after this commit

setotau commented:
I can try this. Also welcome to our matrix chat https://matrix.to/#/#sdm660-mainline:matrix.org

minlexx commented:
> solve gpu crash msm_gem.c

but we don't have any crashes in msm_gem.c. You just commented bunch of stuff that lokks important/useful. And without any explanation as well: which crash, how is commenting calls to drm core is a fix; so it's hard to understand.

ZXlieC commented:
> > solve gpu crash msm_gem.c
> 
> but we don't have any crashes in msm_gem.c. You just commented bunch of stuff that loks important/useful. And without any explanation as well: which crash, how is commenting calls to drm core is a fix; so it's hard to understand.

it solves gpu fault. Like really i don't like to make full description of  commits. At all i'm not supper programer. I just know C syntax and just a little bit how does kernel drivers work. I have tried to copy msm_gem_close function code from 6.16 and it works (i have randomly chosen this function)

minlexx commented:
It's surprisingly all fine on sdm630-sony-pioneer without adjusting `FD_MESA_DEBUG`. Also same on 636 whyred and 660 jasmine. Can consider this all tested then I suppose..

<img width="799" height="1280" alt="image" src="https://github.com/user-attachments/assets/bfb9f167-aabc-42d0-a487-00cbe7f4949e" />

minlexx committed:
dt-bindings: clock: mmcc-sdm660: Add missing MDSS reset

Add definition for display subsystem reset control.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
clk: qcom: mmcc-sdm660: Add missing MDSS reset

Add offset for display subsystem reset in multimedia clock controller
block.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
arm64: dts: qcom: sdm630: Add missing MDSS reset

If the OS does not support recovering the state left by the
bootloader it needs a way to reset display hardware, so that it can
start from a clean state. Add a reference to the relevant reset.

It fixes display init issue appeared in Linux v6.17: without reset
device boots into black screen and you need to turn display off/on
to "fix" it. Also sometimes it can boot into solid blue color
with these messages in kernel log:

  hw recovery is not complete for ctl:2
  [drm:dpu_encoder_phys_vid_prepare_for_kickoff:569] [dpu error]enc33 intf1 ctl 2 reset failure: -22
  [drm:dpu_encoder_frame_done_timeout:2727] [dpu error]enc33 frame done timeout

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>
